res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -1,17 +1,80 @@
 // eslint-disable-next-line-no-var
 var Prime4 = {};
 
+/*
+ * DENON DJ PRIME 4 - CONTROLLER MAPPING FOR MIXXX
+ *
+ * Thank you for using this mapping! Below are a few parameters
+ * you can change if you want to customize anything :)
+ */
+
+// What colour would you like each deck to be?
+// Choose between "red", "yellow", "green", "cyan", "blue", "magenta", or "white".
+const colourOfDeck1 = "white";
+const colourOfDeck2 = "yellow";
+const colourOfDeck3 = "cyan";
+const colourOfDeck4 = "magenta";
+
+/*
+ *                   WARNING!!!
+ *
+ *      DO NOT EDIT ANYTHING PAST THIS POINT
+ *       UNLESS YOU KNOW WHAT YOU'RE DOING.
+ */
+
+Prime4.rgbCode = {
+    black: 0,
+    blueDark: 1,
+    blueDim: 2,
+    blue: 3,
+    greenDark: 4,
+    cyanDark: 5,
+    greenDim: 8,
+    cyanDim: 10,
+    green: 12,
+    cyan: 15,
+    redDark: 16,
+    magentaDark: 17,
+    yellowDark: 20,
+    whiteDark: 21,
+    redDim: 32,
+    magentaDim: 34,
+    yellowDim: 40,
+    whiteDim: 42,
+    red: 48,
+    magenta: 51,
+    orange: 56,
+    yellow: 60,
+    white: 63,
+};
+
+const colDeck1 = eval("Prime4.rgbCode." + colourOfDeck1);
+const colDeck2 = eval("Prime4.rgbCode." + colourOfDeck2);
+const colDeck3 = eval("Prime4.rgbCode." + colourOfDeck3);
+const colDeck4 = eval("Prime4.rgbCode." + colourOfDeck4);
+const colDeck1Dark = eval("Prime4.rgbCode." + colourOfDeck1 + "Dark");
+const colDeck2Dark = eval("Prime4.rgbCode." + colourOfDeck2 + "Dark");
+const colDeck3Dark = eval("Prime4.rgbCode." + colourOfDeck3 + "Dark");
+const colDeck4Dark = eval("Prime4.rgbCode." + colourOfDeck4 + "Dark");
+
 Prime4.init = function() {
     midi.sendShortMsg(0x90, 0x75, 0x00); // Turn off all LEDs
     Prime4.leftDeck = new Prime4.Deck([1, 3], 4);
     Prime4.rightDeck = new Prime4.Deck([2, 4], 5);
     components.Button.prototype.isPress = function(channel, control, value, status) {
         return (status & 0xF0) === 0x90;
     };
-    midi.sendShortMsg(0x9F, 0x1C, 0x7F);
-    midi.sendShortMsg(0x9F, 0x1D, 0x7F);
-    midi.sendShortMsg(0x9F, 0x1E, 0x05);
-    midi.sendShortMsg(0x9F, 0x1F, 0x05);
+    midi.sendShortMsg(0x90, 0x0D, colDeck1Dark); // PFL 1
+    midi.sendShortMsg(0x91, 0x0D, colDeck2Dark); // PFL 2
+    midi.sendShortMsg(0x92, 0x0D, colDeck3Dark); // PFL 3
+    midi.sendShortMsg(0x93, 0x0D, colDeck4Dark); // PFL 4
+    midi.sendShortMsg(0x9F, 0x1C, colDeck1); // Deck 1 Toggle
+    midi.sendShortMsg(0x9F, 0x1D, colDeck2); // Deck 2 Toggle
+    midi.sendShortMsg(0x9F, 0x1E, colDeck3Dark); // Deck 3 Toggle
+    midi.sendShortMsg(0x9F, 0x1F, colDeck4Dark); // Deck 4 Toggle
+    midi.sendShortMsg(0x94, 0x21, colDeck1); // Left Jog Wheel
+    midi.sendShortMsg(0x95, 0x21, colDeck2); // Right Jog Wheel
+    midi.sendShortMsg(0x9F, 0x07, 0x01); // View Button
```

Comment:
most of these are handled automatically by componentsJS if you use it correctly.

Reply:
I have yet to figure out how to use ComponentsJS 'correctly' in this sense, but as soon as I figure this all out I'll make sure to remove these.

Reply:
basically once the controller is fully set up, it will call `Component.trigger()` on each component in your mapping. That will invoke each CO callback connected to each controller once. So thats how the controller gets to sync its leds with mixxx. In the cases where there is no CO associated with the button (such as with the deck toggle buttons), you simply overwrite the `trigger()` method and provide your own "initilization" code in there. This way we properly use OO principles and handle everything related to that physical button the controller in the same software component. When you overwrite the `trigger()` method, you can also use `this.send(value)` instead of specifying the full midi triple, since the component knows the first two bytes (since you passed those using `midi: `).

Reply:
Right, gotcha. I'll have a crack at this when I can and see how I go :)

Reply:
Update: Going to resolve this for now, as I'm unable to cause these LEDs to initialize by themselves, but I'll make sure I try to fix it at a later date. Got more core functionality to sort out first.

Reply:
I could try to implement it for you and send you a PR if you'd like. I took the liberty to unresolve again so I don't forget about it. In the meantime, you're of course free to focus on implementing other features.

Reply:
Yes please, that would be greatly appreciated :)